pkg/limits/ingest_limits.go

```diff
@@ -389,62 +353,57 @@ func (s *IngestLimits) GetAssignedPartitions(_ context.Context, _ *logproto.GetA
-	cutoff := time.Now().Add(-s.cfg.WindowSize).UnixNano()
-
-	// Calculate the rate window cutoff in nanoseconds
-	rateWindowCutoff := time.Now().Add(-s.cfg.RateWindow).UnixNano()
-
-	// Count total active streams for the tenant
-	// across all assigned partitions and record
-	// the streams that have been seen within the
-	// window
+func (s *IngestLimits) ExceedsLimits(_ context.Context, req *logproto.ExceedsLimitsRequest) (*logproto.ExceedsLimitsResponse, error) {
 	var (
-		activeStreams uint64
-		totalSize     uint64
+		lastSeenAt = time.Now()
+		// Use the provided lastSeenAt timestamp as the last seen time
+		recordTime = lastSeenAt.UnixNano()
+		// Calculate the cutoff for the window size
+		cutoff = lastSeenAt.Add(-s.cfg.WindowSize).UnixNano()
+		// Get the bucket for this timestamp using the configured interval duration
+		bucketStart = lastSeenAt.Truncate(s.cfg.BucketDuration).UnixNano()
+		// Calculate the rate window cutoff for cleaning up old buckets
+		bucketCutoff = lastSeenAt.Add(-s.cfg.RateWindow).UnixNano()
+		// Calculate the max active streams per tenant per partition
+		maxActiveStreams = uint64(s.limits.MaxGlobalStreamsPerUser(req.Tenant) / s.cfg.NumPartitions)
+		// Create a map of streams per partition
+		streams = make(map[int32][]Stream)
 	)
 
-	// If the stream is written into a partition we are
-	// assigned to and has been seen within the window,
-	// it is an active stream.
-	unknownStreams := req.StreamHashes
-
-	s.metadata.Usage(req.Tenant, func(partitionID int32, stream Stream) {
-		if assigned := s.partitionManager.Has(partitionID); !assigned {
-			return
-		}
+	for _, stream := range req.Streams {
+		partitionID := int32(stream.StreamHash % uint64(s.cfg.NumPartitions))
 
-		if stream.LastSeenAt < cutoff {
-			return
-		}
+		streams[partitionID] = append(streams[partitionID], Stream{
+			Hash:       stream.StreamHash,
+			LastSeenAt: recordTime,
+			TotalSize:  stream.EntriesSize + stream.StructuredMetadataSize,
+		})
+	}
 
-		activeStreams++
+	exceedLimits, ingestedBytes := s.metadata.StoreIf(req.Tenant, streams, maxActiveStreams, cutoff, bucketStart, bucketCutoff)
 
-		// Calculate size only within the rate window
-		for _, bucket := range stream.RateBuckets {
-			if bucket.Timestamp >= rateWindowCutoff {
-				totalSize += bucket.Size
-			}
+	var results []*logproto.ExceedsLimitsResult
+	for reason, streamHashes := range exceedLimits {
+		for _, streamHash := range streamHashes {
+			results = append(results, &logproto.ExceedsLimitsResult{
+				StreamHash: streamHash,
+				Reason:     uint32(reason),
+			})
 		}
+	}
 
-		for i, streamHash := range unknownStreams {
-			if stream.Hash == streamHash {
-				unknownStreams = append(unknownStreams[:i], unknownStreams[i+1:]...)
-				break
-			}
-		}
-	})
-
-	// Calculate rate using only data from within the rate window
-	rate := float64(totalSize) / s.cfg.RateWindow.Seconds()
+	s.metrics.tenantIngestedBytesTotal.WithLabelValues(req.Tenant).Add(float64(ingestedBytes))
```

Comment:
I think it's OK to move this metric to `StoreIf`

Reply:
My preference is to not overload the stream-metadata implementation with more dependencies than needed. Metrics are relevant to the hosting service. WDYT?